File: 3750-closest-equal-element-queries/solution.java
import java.util.*;

class Solution {
    public List<Integer> solveQueries(int[] nums, int[] queries) {

        int n = nums.length;

        
        Map<Integer, List<Integer>> map = new HashMap<>();

        for (int i = 0; i < n; i++) {
            map.computeIfAbsent(nums[i], k -> new ArrayList<>()).add(i);
        }

        List<Integer> ans = new ArrayList<>();

        for (int q : queries) {

            List<Integer> list = map.get(nums[q]);

            
            if (list.size() == 1) {
                ans.add(-1);
                continue;
            }

            int pos = Collections.binarySearch(list, q);

            int leftIdx = list.get((pos - 1 + list.size()) % list.size());
            int rightIdx = list.get((pos + 1) % list.size());

            int d1 = Math.abs(q - leftIdx);
            int d2 = Math.abs(q - rightIdx);

            int dist1 = Math.min(d1, n - d1);
            int dist2 = Math.min(d2, n - d2);

            ans.add(Math.min(dist1, dist2));
        }

        return ans;
    }
}

File: 3750-closest-equal-element-queries/solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> solveQueries(vector<int>& nums, vector<int>& queries) {
        int n = nums.size();

       
        unordered_map<int, vector<int>> mp;

        for (int i = 0; i < n; i++) {
            mp[nums[i]].push_back(i);
        }

        vector<int> ans;

        for (int q : queries) {
            vector<int>& list = mp[nums[q]];

            // if only one occurrence
            if (list.size() == 1) {
                ans.push_back(-1);
                continue;
            }

            int pos = lower_bound(list.begin(), list.end(), q) - list.begin();

            int size = list.size();

            int leftIdx = list[(pos - 1 + size) % size];
            int rightIdx = list[(pos + 1) % size];

            int d1 = abs(q - leftIdx);
            int d2 = abs(q - rightIdx);

            int dist1 = min(d1, n - d1);
            int dist2 = min(d2, n - d2);

            ans.push_back(min(dist1, dist2));
        }

        return ans;
    }
};
